File: csharp/src/LabelSuite.Core/Report.cs
// LOT별 검사 리포트 xlsx 내보내기 (파이썬 core/history/report.py 포팅).
using ClosedXML.Excel;

namespace LabelSuite.Core;

public static class Report
{
    public static int ExportLotReport(HistoryDb db, string lot, string outPath)
    {
        var rows = db.Query(lot: lot, limit: 10_000);
        using var workbook = new XLWorkbook();

        var summary = workbook.AddWorksheet("요약");
        var passedCount = rows.Count(r => r.Passed);
        var summaryRows = new (string, string)[]
        {
            ("LOT", lot),
            ("검사 건수", rows.Count.ToString()),
            ("합격", passedCount.ToString()),
            ("확인 필요", (rows.Count - passedCount).ToString()),
            ("합격률", rows.Count > 0 ? $"{passedCount * 100.0 / rows.Count:F1}%" : "-"),
        };
        for (var i = 0; i < summaryRows.Length; i++)
        {
            summary.Cell(i + 1, 1).SetValue(summaryRows[i].Item1).Style.Font.SetBold();
            summary.Cell(i + 1, 2).SetValue(summaryRows[i].Item2);
        }

        var detail = workbook.AddWorksheet("검사 상세");
        var header = new[] { "일시", "LOT", "REF", "PN", "규격", "소스", "페이지",
                             "판정", "이미지 경로", "필드별 검출/기준" };
        for (var i = 0; i < header.Length; i++)
            detail.Cell(1, i + 1).SetValue(header[i]).Style.Font.SetBold();
        var rowNumber = 2;
        foreach (var row in rows)
        {
            var fields = db.FieldsFor(row.Id);
            var fieldSummary = string.Join(", ", fields
                .Where(f => f.Expected is not null)
                .Select(f => $"{f.Field} {f.Found}/{f.Expected}"));
            var values = new[]
            {
                row.Ts, row.Lot, row.Ref, row.Pn, row.Standard, row.Source,
                row.Page is { } p ? (p + 1).ToString() : "",
                row.Passed ? "합격" : "확인 필요", row.ImagePath, fieldSummary,
            };
            for (var i = 0; i < values.Length; i++)
                detail.Cell(rowNumber, i + 1).SetValue(values[i]);
            detail.Cell(rowNumber, 8).Style.Fill.SetBackgroundColor(
                row.Passed ? XLColor.FromArgb(0x90, 0xEE, 0x90)
                           : XLColor.FromArgb(0xFF, 0xE4, 0xB5));
            rowNumber++;
        }
        summary.Columns().AdjustToContents();
        detail.Columns().AdjustToContents();
        workbook.SaveAs(outPath);
        return rows.Count;
    }
}

File: labelsuite/core/history/report.py
"""LOT별 검사 리포트 내보내기 (xlsx)."""

from __future__ import annotations

from openpyxl import Workbook
from openpyxl.styles import Font, PatternFill

from labelsuite.core.history.db import HistoryDb

_PASS_FILL = PatternFill("solid", start_color="90EE90")
_FAIL_FILL = PatternFill("solid", start_color="FFE4B5")
_BOLD = Font(bold=True)


def export_lot_report(db: HistoryDb, lot: str, out_path: str) -> int:
    """지정 LOT의 검사 이력을 요약+상세 시트로 내보낸다. 반환: 검사 건수."""
    rows = db.query(lot=lot, limit=10_000)
    wb = Workbook()

    summary = wb.active
    summary.title = "요약"
    summary.append(["LOT", lot])
    summary.append(["검사 건수", len(rows)])
    passed_count = sum(1 for r in rows if r.passed)
    summary.append(["합격", passed_count])
    summary.append(["확인 필요", len(rows) - passed_count])
    summary.append(["합격률",
                    f"{passed_count / len(rows) * 100:.1f}%" if rows else "-"])
    for row in summary.iter_rows(max_row=5, max_col=1):
        row[0].font = _BOLD

    detail = wb.create_sheet("검사 상세")
    header = ["일시", "LOT", "REF", "PN", "규격", "소스", "페이지", "판정",
              "이미지 경로", "필드별 검출/기준"]
    detail.append(header)
    for cell in detail[1]:
        cell.font = _BOLD
    for row in rows:
        fields = db.fields_for(row.id)
        field_summary = ", ".join(
            f"{f.field} {f.found}/{f.expected}" for f in fields
            if f.expected is not None)
        detail.append([
            row.ts, row.lot, row.ref, row.pn, row.standard, row.source,
            (row.page + 1) if row.page is not None else "",
            "합격" if row.passed else "확인 필요",
            row.image_path, field_summary])
        verdict_cell = detail.cell(row=detail.max_row, column=8)
        verdict_cell.fill = _PASS_FILL if row.passed else _FAIL_FILL

    for sheet in (summary, detail):
        for col_cells in sheet.columns:
            width = max(len(str(c.value or "")) for c in col_cells)
            sheet.column_dimensions[col_cells[0].column_letter].width = min(width + 2, 50)

    wb.save(out_path)
    return len(rows)
